// ent42/abstractCubeAiBot.js
const { program } = require('commander');
const { ContextCommands } = require('./contextCommands.js');
const { AIInterface } = require('../../AIInterface.js');
const { Licensing } = require("ent42/license.js");
const { sleep, debugOut, writeFileSafely } = require('../../vxAssistCommon.js');
const fs = require('fs');

class AbstractCubeAiBot {
  constructor() {
    this.packageJson = require('./package.json');
    this.storageFilePath = './db/botStorage.json';
    this.cachesFilePath = './db/botCaches.json';

    this.adminUsers = [];

    this.commands = new ContextCommands();

    this.ai = {};
    this.aiIgnoreReply = {};

    this.aiParams = {};
    this.aiContext = {};

    this.caches = {
      users: {},
      groups: {}
    };

    Licensing.isAdmin = (userId) => {
      return this.isBotAdmin(userId)
    };
  }

  loadFromStorage() {
    if (fs.existsSync(this.storageFilePath)) {
      const data = fs.readFileSync(this.storageFilePath, 'utf8');
      const storage = JSON.parse(data);
      this.botToken = storage.botToken;
      this.adminUsers = storage.adminUsers;
      this.aiParams = storage.aiParams ?? {};
      this.aiContext = storage.aiContext ?? {}
    }

    if (fs.existsSync(this.cachesFilePath)) {
      try {
        this.caches = JSON.parse(fs.readFileSync(this.cachesFilePath, 'utf8'));
      } catch (error) {
        this.caches = {
          users: {},
          groups: {}
        }
        console.error(error);
      }
    }
  }

  saveToStorage() {
    debugOut(`Saving data to ${this.storageFilePath}`);

    const purgeOrphanConfigEntries = (config) => {
      const defaultConfig = this.getDefaultConfig();
      const newConfig = { ...config }

      Object.keys(newConfig).forEach(i => {
        if (!defaultConfig.hasOwnProperty(i)) {
          delete newConfig[i];
        }
      });

      return newConfig;
    }

    if (this.ai) {
      Object.keys(this.ai).forEach(i => {
        Object.keys(this.ai[i]).forEach(j => {
          // collect local AI's configuration from runtime
          if (!this.aiParams[i]) { this.aiParams[i] = {}; }
          this.aiParams[i][j] = purgeOrphanConfigEntries(this.ai[i][j].config);
          // collect local AI's context from runtime
          if (!this.aiContext[i]) { this.aiContext[i] = {}; }
          this.aiContext[i][j] = [...this.ai[i][j].uniqueAi.messages];
        })
      });
    }

    const storage = {
      botToken: this.botToken ?? "",
      adminUsers: this.adminUsers ?? "",
      aiParams: this.aiParams ?? "",
      aiContext: this.aiContext ?? ""
    };

    writeFileSafely(this.storageFilePath, JSON.stringify(storage, null, 2), 'utf-8');
    writeFileSafely(this.cachesFilePath, JSON.stringify(this.caches, null, 2), 'utf-8');
  }

  isBotAdmin(userId) {
    return this.adminUsers.includes(userId) || this.isBotOwner(userId);
  }

  isBotOwner(userId) {
    return userId === this.getOwner();
  }

  getOwner() {
    return 271199477;
  }

  async SetupContextCommands() {
    throw new Error('Not implemented');
  }

  async onRun() {
    throw new Error('Not implemented');
  }

  async run() {
    const cmdLine = parseCommandLine();

    if (cmdLine.delay) {
      /** 
       * Delaying the startup was neccessary on macOS / launchd due to 
       * the murus firewall interrupting polling for unknown reasons.
      */
      debugOut('Startup delay: ' + cmdLine.delay + ' seconds ...');
      await sleep(cmdLine.delay * 1000);
    }

    setupSignalHandler();

    this.onRun({ cmdLine });
  }

  async onShutdown() {
    throw new Error('Not implemented');
  }

  async shutdown() {
    try {
      debugOut('Bot shutting down...');
      this.alertOwner('Bot shutdown');
    } catch (ex) {
      debugOut(`${ex.message}\n${ex.stack}`);
    }

    return this.onShutdown().catch(ex => {
      debugOut(`${ex.message}\n${ex.stack}`);
    }).finally(() => {

      try {
        Licensing.saveToStorage();
      } catch (ex) {
        debugOut(`${ex.message}\n${ex.stack}`);
      }
  
      try {
        this.saveToStorage();
      } catch (ex) {
        debugOut(`${ex.message}\n${ex.stack}`);
      }

      debugOut('Bot stopped');
      process.exit(0);
    });
  }
}

const setupSignalHandler = () => {
  process.on('SIGINT', () => { debugOut('!!! SIGINT !!!'); this.shutdown(); });
  process.on('SIGTERM', () => { debugOut('!!! SIGTERM !!!'); this.shutdown(); });
}

const parseCommandLine = () => {
  program
    .option('-d, --delay <param>', 'startup delay in seconds', parseInt)
    .parse(process.argv);

  return program.opts();
}

module.exports = {
  AbstractCubeAiBot
};
