// ent42/abstractCubeAiBot.js
const { program } = require('commander');
const { ContextCommands } = require('./contextCommands.js');
const { AIInterface } = require('../../AIInterface.js');
const { Licensing } = require("ent42/license.js");
const { sleep, debugOut } = require('../../vxAssistCommon.js');

class AbstractCubeAiBot {
    constructor() {
        this.packageJson = require('./package.json');
        this.storageFilePath = './db/botStorage.json';
        this.cachesFilePath = './db/botCaches.json';

        this.adminUsers = [];

        this.commands = new ContextCommands();

        this.ai = {};
        this.aiIgnoreReply = {};

        this.aiParams = {};
        this.aiContext = {};

        this.caches = {
            users: {},
            groups: {}
        };

        Licensing.isAdmin = (userId) => {
            return this.isBotAdmin(userId)
        };
    }

    isBotAdmin(userId) {
        return this.adminUsers.includes(userId) || this.isBotOwner(userId);
    }

    isBotOwner(userId) {
        return userId === this.getOwner();
    }

    getOwner() {
        return 271199477;
    }

    async SetupContextCommands() {
        throw new Error('Not implemented');
    }

    async onRun() {
        throw new Error('Not implemented');
    }

    async run() {
        const cmdLine = parseCommandLine();

        if (cmdLine.delay) {
            /** 
             * Delaying the startup was neccessary on macOS / launchd due to 
             * the murus firewall interrupting polling for unknown reasons.
            */
            debugOut('Startup delay: ' + cmdLine.delay + ' seconds ...');
            await sleep(cmdLine.delay * 1000);
        }

        setupSignalHandler();

        this.onRun({ cmdLine });
    }

    async onShutdown() {
        throw new Error('Not implemented');
    }

    async shutdown() {
        try {
            debugOut('Bot shutting down...');
            this.alertOwner('Bot shutdown');
        } catch (ex) {
            debugOut(`${ex.message}\n${ex.stack}`);
        }

        try {
            Licensing.saveToStorage();
        } catch (ex) {
            debugOut(`${ex.message}\n${ex.stack}`);
        }

        try {
            this.saveToStorage();
        } catch (ex) {
            debugOut(`${ex.message}\n${ex.stack}`);
        }

        return this.onShutdown().catch(ex => {
            debugOut(`${ex.message}\n${ex.stack}`);
        }).finally(() => {
            debugOut('Bot stopped');
            process.exit(0);
        });
    }
}

const setupSignalHandler = () => {
    process.on('SIGINT', () => { debugOut('!!! SIGINT !!!'); this.shutdown(); });
    process.on('SIGTERM', () => { debugOut('!!! SIGTERM !!!'); this.shutdown(); });
}

const parseCommandLine = () => {
    program
        .option('-d, --delay <param>', 'startup delay in seconds', parseInt)
        .parse(process.argv);

    return program.opts();
}

module.exports = {
    AbstractCubeAiBot
};
