// ent42/contextCommands.js
class ContextCommands {
    constructor() {
        this.mapCommand2Callback = {};
    }

    addBotOwner(command, callback, description) {
        description = '[Bot Owner] ' + description;
        return this.registerCommand(command, callback, description, true, false, 'BotOwner');
    }

    addBotAdmin(command, callback, description) {
        description = '[Bot Admin] ' + description;
        return this.registerCommand(command, callback, description, false, true, 'BotAdmin');
    }

    addGroupAdmin(command, callback, description) {
        description = '[Group Admin] ' + description;
        return this.registerCommand(command, callback, description, false, false, 'GroupAdmin');
    }

    addGroup(command, callback, description) {
        description = '[Group] ' + description;
        return this.registerCommand(command, callback, description, false, false, 'Group');
    }

    addUser(command, callback, description) {
        description = '[User] ' + description;
        return this.registerCommand(command, callback, description, false, false, 'User');
    }

    registerCommand(command, callback, description, ownerOnly, adminOnly, scope) {
        scope = this.translateScope(scope);        
        this.mapCommand2Callback[scope] ??= {};
        this.mapCommand2Callback[scope][command] = {
            callback,
            ownerOnly,
            adminOnly,
            description,
            scope
        }

        return this.mapCommand2Callback[scope][command];
    }

    translateScope(localScope) {
        switch (localScope) {
            case 'BotOwner': return '%BotOwner%';
            case 'BotAdmin': return '%BotAdmin%';
            case 'User': return 'all_private_chats'
            case 'GroupAdmin': return 'all_chat_administrators';
            case 'Group': return 'all_group_chats';
            default: return 'default'
        }
    }

    async forEachUniqueCommand(callback) {
        var uniqueCommands = {};

        Object.keys(this.mapCommand2Callback).forEach(scope => {
            uniqueCommands = { ...uniqueCommands, ...this.mapCommand2Callback[scope] };
        });

        Object.keys(uniqueCommands).forEach(i => {
            callback(i, uniqueCommands[i]);
        });
    }

    async forEachScope(callback) {
        Object.keys(this.mapCommand2Callback).forEach(scope => {
            let commands = Object.keys(this.mapCommand2Callback[scope]).map((command) => {
                const c = this.mapCommand2Callback[scope][command];
                return { command, description: c.description, scope: c.scope };
            });
            callback(scope, commands);
        });
    }

    getByName(command) {
        for (let scope of Object.values(this.mapCommand2Callback)) {
            if (scope[command]) {
                return scope[command];
            }
        }

        return null;
    }
}

module.exports = {
    ContextCommands
};
