// ent42/license.js
const { debugOut } = require('../../vxAssistCommon.js');
const { v4: uuidv4 } = require('uuid');
const fs = require('fs');

const LicenseType = {
    Private: 'private',
    Group: 'group'
}

function defaultLicense() {
    return {
        licId: uuidv4(),
        type: LicenseType.Group,
        revoked: false,
        shareable: false,        
        consumer: null,
        expiration: Date.now() + 1000 * 60 * 60 * 24 * 365,

        modules: {
            openai: {
                models: ['gpt-3.5-turbo', 'gpt-3.5-turbo-16k'],
                limits: {
                    tokenUsed: 0,
                    tokenMax: 10 * 4096,
                }
            },
            huggingface: {
                models: ['dreamlike-art/dreamlike-anime-1.0', 'dreamlike-art/dreamlike-diffusion-1.0'],
                limits: {
                    tokenUsed: 0,
                    tokenMax: 10 * 4096,
                }
            },
        },
    };
}

class Licensing {
    constructor() {
        this.storageFilePath = 'botLicenses.json';
        this.loadFromStorage();        
    }

    revoke(license) {
        license.revoked = true;

        this.licenses.forEach((l) => {
            if (l.licId === license.licId) {
                l.revoked = true;
            }
        });

        this.saveToStorage();
    }

    issue() {
        const license = defaultLicense();
        this.licenses.push(license);
        this.saveToStorage();

        return license;
    }

    claim(licId, consumer) {
        var license = this.getById(licId);
        var current = this.getByConsumer(consumer);

        if(!consumer) { 
            /** consumber must be specified 
             */
            return false; 
        }

        if(!license) { 
            debugOut(`License key ${licId} is invalid`);
            return false; 
        }

        if(license.revoked) { 
            debugOut(`License key ${licId} is revoked`);
            return false; 
        }

        if (license.consumer && license.consumer === consumer) {
            debugOut(`License key ${licId} is already claimed by ${consumer}`);
            return false;
        }

        if (!license.shareable && license.consumer !== null && license.consumer !== consumer) {
            debugOut(`Unshareable License key ${licId} for ${consumer} is already claimed by ${license.consumer}`);
            return false;
        } 
        
        if(license.shareable && license.consumer !== null && license.consumer !== consumer){
            debugOut(`Shareable license key ${licId} will be transfered form ${license.consumer} to ${consumer}`);
        }
        
        if(current){
            debugOut(`Current license ${current.licId} for ${consumer} replaced by and linked to ${licId}`);
            this.licenses = this.licenses.filter((l) => l.licId !== current.licId);
            license.previousLicense = current;
        }

        license.consumer = consumer;            
        this.saveToStorage();

        debugOut(`License key ${licId} claimed by ${consumer}`);

        return true;
    }

    validate(license) {
        return !license.revoked;
    }

    getByConsumer(consumer) {
        return this.licenses.find((l) => l.consumer === consumer);

    }

    getById(licId) {
        return this.licenses.find((l) => l.licId === licId);
    }

    saveToStorage() {
        debugOut(`Saving licenses to ${this.storageFilePath}`);

        var dbo = {
            lastChanged: new Date().toISOString(),
            licenses: this.licenses
        }

        fs.writeFileSync('licenseStore.json', JSON.stringify(dbo, null, 2), 'utf8');
    }

    loadFromStorage() {
        debugOut(`Loading licenses from ${this.storageFilePath}`);

        if(fs.existsSync(this.storageFilePath)) {
            var dbo = JSON.parse(fs.readFileSync(this.storageFilePath, 'utf8'));
            this.licenses = dbo.licenses;
        }else{
            this.licenses = [];
        }

        return this.licenses;
    }
}

module.exports = {
    Licensing: new Licensing,
    LicenseType: LicenseType
};
