// ent42/license.js
const { debugOut } = require('../../vxAssistCommon.js');
const { v4: uuidv4 } = require('uuid');
const fs = require('fs');
const { group } = require('console');

const LicenseType = {
    Private: 'private',
    Group: 'group'
}

/**
 * 4K context	$0.0015 / 1K tokens	$0.002 / 1K tokens
 * 16K context	$0.003 / 1K tokens	$0.004 / 1K tokenss
*/

class Licensing {
    constructor() {
        this.storageFilePath = 'botLicenses.json';
        this.loadFromStorage();
        this.isAdmin = (userId) => false;
    }

    getDefaultLicense() {
        return {
            licId: uuidv4(),
            type: LicenseType.Group,
            revoked: false,
            shareable: false,
            consumer: null,
            unresticted: false,
            expiration: Date.now() + 1000 * 60 * 60 * 24 * 365,

            modules: {
                openai: {
                    models: ['gpt-3.5-turbo', 'gpt-3.5-turbo-16k'],
                    limits: {
                        tokenUsed: 0,
                        tokenMax: 250000,
                    }
                },
                huggingface: {
                    models: ['dreamlike-art/dreamlike-anime-1.0', 'dreamlike-art/dreamlike-diffusion-1.0'],
                    limits: {
                        tokenUsed: 0,
                        tokenMax: 25,
                    }
                },
            },
        };
    }

    revoke(license) {
        license.revoked = true;

        this.licenses.forEach((l) => {
            if (l.licId === license.licId) {
                l.revoked = true;
            }
        });

        this.saveToStorage();
    }

    issue() {
        const license = this.getDefaultLicense();
        this.licenses.push(license);
        this.saveToStorage();

        return license;
    }

    claim(licId, consumer) {
        var license = this.getById(licId);
        var current = this.getByConsumer(consumer);

        if (!consumer) {
            /** consumber must be specified 
             */
            return false;
        }

        if (!license) {
            debugOut(`License key ${licId} is invalid`);
            return false;
        }

        if (license.revoked) {
            debugOut(`License key ${licId} is revoked`);
            return false;
        }

        if (license.consumer && license.consumer === consumer) {
            debugOut(`License key ${licId} is already claimed by ${consumer}`);
            return false;
        }

        if (!license.shareable && license.consumer !== null && license.consumer !== consumer) {
            debugOut(`Unshareable License key ${licId} for ${consumer} is already claimed by ${license.consumer}`);
            return false;
        }

        if (license.shareable && license.consumer !== null && license.consumer !== consumer) {
            debugOut(`Shareable license key ${licId} will be transfered form ${license.consumer} to ${consumer}`);
        }

        if (current) {
            debugOut(`Current license ${current.licId} for ${consumer} replaced by and linked to ${licId}`);
            this.licenses = this.licenses.filter((l) => l.licId !== current.licId);
            license.previousLicense = current;
            
            Object.keys(this.groups).forEach((i) => {
                const g = this.groups[i];

                if(g.licId === current.licId){
                    g.licId = licId;
                }
            });
        }

        license.consumer = consumer;
        this.saveToStorage();

        debugOut(`License key ${licId} claimed by ${consumer}`);

        return true;
    }

    validate(license) {
        if (!license) {
            return false;
        }

        if (license.revoked !== false) {
            return false;
        }

        return true
    }

    getByConsumer(consumer) {
        var license = this.licenses.find((l) => l.consumer === consumer);

        if (!license && this.isAdmin(consumer)) {
            license = this.getRootLicense();
        }

        return license;
    }

    getById(licId) {
        if (licId === this.getRootLicense().licId) {
            return this.getRootLicense();
        }
        return this.licenses.find((l) => l.licId === licId);
    }

    getByGroupId(groupId) {
        return this.getById(this.groups[groupId]);
    }

    getRootLicense() {
        var licence = this.getDefaultLicense();

        licence.licId = '%ADMINLICENSE%';
        licence.shareable = false;
        licence.unresticted = true;
        licence.consumer = 'root';

        return licence;
    }

    registerGroup(groupId, license) {
        if (!this.validate(license)) {
            throw new Error(`License invalid`);
        }

        if (license.type !== LicenseType.Group) {
            throw new Error(`License invalid: single user license`);
        }

        if (this.groups[groupId] && this.groups[groupId] === license.licId) {
            throw new Error(`License already registered`);
        }

        if (this.groups[groupId] && this.groups[groupId] !== license.licId) {
            throw new Error(`License already registered`);
        }

        this.groups[groupId] = license.licId;
        this.saveToStorage();

        debugOut(`Group ${groupId} registered with license ${license.licId}`);
    }

    deregisterGroup(groupId, license) {
        if (!license) {
            throw new Error('License is null');
        }

        if (!this.groups[groupId]) {
            throw new Error('Group has no license registereds');
        }

        if (this.groups[groupId] !== license.licId) {
            throw new Error('Group is registered with a different license');
        }

        delete this.groups[groupId];
        this.saveToStorage();
        debugOut(`Group ${groupId} deregistered with license ${license.licId}`);
    }

    saveToStorage() {
        debugOut(`Saving licenses to ${this.storageFilePath}`);

        var dbo = {
            lastChanged: new Date().toISOString(),
            licenses: this.licenses,
            groups: this.groups
        }

        fs.writeFileSync(this.storageFilePath, JSON.stringify(dbo, null, 2), 'utf8');
    }

    loadFromStorage() {
        debugOut(`Loading licenses from ${this.storageFilePath}`);

        if (fs.existsSync(this.storageFilePath)) {
            var dbo = JSON.parse(fs.readFileSync(this.storageFilePath, 'utf8'));
            this.licenses = dbo.licenses ?? [];
            this.groups = dbo.groups ?? {};
        } else {
            this.licenses = [];
            this.groups = {};
        }

        return this.licenses;
    }
}

module.exports = {
    Licensing: new Licensing,
    LicenseType: LicenseType
};
