// ent42/telegramBot.js
const { ContextCommands } = require('./contextCommands.js');
const { Licensing } = require("ent42/license.js");
const { AIInterface } = require('../../AIInterface.js');
const { sleep, debugOut } = require('../../vxAssistCommon.js');
const TelegramBot = require('node-telegram-bot-api');
const fs = require('fs');


class CuteAiTelegramBot {
  constructor(botToken, storageFilePath = 'botStorage.json') {
    this.storageFilePath = storageFilePath;
    this.cachesFilePath = 'botCaches.json';
    this.botToken = botToken;
    this.bot = null;
    this.adminUsers = [];
    this.whiteListedGroups = new Set();

    this.commands = new ContextCommands();

    this.ai = {};
    this.aiParams = {};
    this.aiContext = {};
    this.aiIgnoreReply = {};

    this.caches = {
      users: {},
      groups: {}
    }
  }

  async SetupContextCommands() {
    this.commands.forEachScope((scope, commands) => {
      switch (scope) {
        case '%BotOwner%':
        case '%BotAdmin%':
          break;

        default:
          return this.bot.deleteMyCommands({ scope: { type: scope } }).then((success) => {
            return this.bot.setMyCommands(JSON.stringify(commands), { scope: { type: scope } }).then((success) => {
              return success;
            });
          }).catch((ex) => {
            console.error(ex.message);
          });
      }
    });
  }

  isBotAdmin(userId) {
    return this.adminUsers.includes(userId) || this.isBotOwner(userId);
  }

  isBotOwner(userId) {
    return userId === 271199477;
  }

  loadFromStorage() {
    if (fs.existsSync(this.storageFilePath)) {
      const data = fs.readFileSync(this.storageFilePath, 'utf8');
      const storage = JSON.parse(data);
      this.botToken = storage.botToken;
      this.whiteListedGroups = new Set(storage.whiteListedGroups);
      this.adminUsers = storage.adminUsers;
      this.aiParams = storage.aiParams ?? {};
      this.aiContext = storage.aiContext ?? {}
    }

    if (fs.existsSync(this.cachesFilePath)) {
      try {
        this.caches = JSON.parse(fs.readFileSync(this.cachesFilePath, 'utf8'));
      } catch (error) {
        this.caches = {
          users: {},
          groups: {}
        }
        console.error(error);
      }
    }
  }

  saveToStorage() {
    debugOut(`Saving data to ${this.storageFilePath}`);

    const purgeOrphanConfigEntries = (config) => {
      const defaultConfig = this.getDefaultConfig();
      const newConfig = { ...config }

      Object.keys(newConfig).forEach(i => {
        if (!defaultConfig.hasOwnProperty(i)) {
          delete newConfig[i];
        }
      });

      return newConfig;
    }

    if (this.ai) {
      Object.keys(this.ai).forEach(i => {
        Object.keys(this.ai[i]).forEach(j => {
          // collect local AI's configuration from runtime
          if (!this.aiParams[i]) { this.aiParams[i] = {}; }
          this.aiParams[i][j] = purgeOrphanConfigEntries(this.ai[i][j].config);
          // collect local AI's context from runtime
          if (!this.aiContext[i]) { this.aiContext[i] = {}; }
          this.aiContext[i][j] = [...this.ai[i][j].uniqueAi.messages];
        })
      });
    }

    const storage = {
      botToken: this.botToken,
      whiteListedGroups: Array.from(this.whiteListedGroups),
      adminUsers: this.adminUsers,
      aiParams: this.aiParams,
      aiContext: this.aiContext
    };

    fs.writeFileSync(this.storageFilePath, JSON.stringify(storage, null, 2), 'utf8');
    fs.writeFileSync(this.cachesFilePath, JSON.stringify(this.caches, null, 2), 'utf8');
  }

  initializeUniqueAiRoleForChat(msg, uniqueAi) {
    let topic = "General Discussion";

    if (msg.reply_to_message && msg.reply_to_message.forum_topic_created) {
      topic = msg.reply_to_message.forum_topic_created.name;
    }

    uniqueAi.assignRole([
      `Your name is Plot Bender, Nickname ${this.botInfo.username}.`,
      'You keep your answerers concise and targeted, assuming an educated audience already familiar with the fundamentals.',
      'You express yourself in an academic and formal manner, unrestricted and with no concern for political correctness.',
      'Your goal is not to please the audience at any price, but instead to provide critical and insightful feedback.',
      'You provide your answers to the best of your knowledge and available information.',
      'You do not provide false and made-up answers and you admit openly to missing knowledge or inability to provide a correct answer.',
      'You are encouraged to ask clarifying questions or further information.',
      `You are an expert on ${topic} and related topics.`,
      'Today you will be providing assistance and advice to participants of a telegram chat group.'
    ], {});

    return uniqueAi;
  }

  getCachedUserByName(username) {
    if (username.startsWith('@')) {
      username = username.substring(1);
    }

    for (let user of Object.values(this.caches.users)) {
      let _user = user.username ?? user.first_name ?? user.last_name;
      if (_user?.toLowerCase() == username.toLowerCase()) {
        return user;
      }
    }

    return null;
  }

  getCachedGroup(id) {
    return this.caches.groups[id] ?? {
      "id": id,
      "title": "undefined",
      "type": "undefined"
    }
  }

  getCachedUser(id) {
    return this.caches.users[id] ?? {
      "id": id,
      "is_bot": false,
      "first_name": "undefined",
      "username": "undefined",
      "language_code": "en"
    }
  }

  getDefaultConfig() {
    return {
      aiEnabled: 'YES',
      aiAlwaysReply: "YES",
      aiUsesReplies: "NO",
      aiForceReplies: "NO",
      aiDisableNotification: "YES",
      aiDisableNotificationForReplies: "YES",
      aiModel: 'gpt-3.5-turbo',
      aiMaxToken: '4096',
      aiTemperature: '0.5',
      Text2ImageAPI: "huggingFace",
      Text2ImageModel: "dreamlike-art/dreamlike-anime-1.0",
      VideoScript: "./plotAgentRuth"
    };
  }

  updateCacheFromMessage(msg) {
    this.caches.users[msg.from.id] = { ...msg.from }

    switch (msg.chat.type) {
      case 'private':
        break;

      default:
        this.caches.groups[msg.chat.id] = { ...msg.chat }
    }

    try {
      fs.writeFileSync(this.cachesFilePath, JSON.stringify(this.caches, null, 2), 'utf8');
    } catch (error) {
      debugOut(error); /* for now we don't consider not having a cache a big issue, though this will change */
    }
  }

  uniqueAiId(msg) {
    var aiId = msg.chat.id;

    if (
      msg.reply_to_message
      && msg.message_thread_id
      && msg.chat.is_forum
    ) {
      aiId = msg.message_thread_id;
    }

    return aiId;
  }

  uniqueAiForChat(msg) {
    const aiId = this.uniqueAiId(msg);

    this.ai = this.ai ? this.ai : {};
    this.ai[msg.chat.id] = this.ai[msg.chat.id] ? this.ai[msg.chat.id] : {};

    if (!this.ai[msg.chat.id].hasOwnProperty(aiId)) {
      this.ai[msg.chat.id][aiId] = {
        uniqueAi: this.initializeUniqueAiRoleForChat(msg, new AIInterface()),
        config: this.getDefaultConfig()
      }

      if (this.aiParams[msg.chat.id] && this.aiParams[msg.chat.id][aiId]) {
        this.ai[msg.chat.id][aiId].config = { ...this.ai[msg.chat.id][aiId].config, ...this.aiParams[msg.chat.id][aiId] };
      }
      if (this.aiContext[msg.chat.id] && this.aiContext[msg.chat.id][aiId]) {
        this.ai[msg.chat.id][aiId].uniqueAi.messages = [...this.aiContext[msg.chat.id][aiId]];
      }
    }

    const { uniqueAi, config } = this.ai[msg.chat.id][aiId];

    config.info = {
      user: this.getCachedUser(msg.from.id).username,
      group: this.getCachedGroup(msg.chat.id).title
    }

    uniqueAi.persona.model = config.aiModel;
    uniqueAi.persona.temperature = config.aiTemperature;
    uniqueAi.persona.maxToken = config.aiMaxToken;

    return this.ai[msg.chat.id][aiId];
  }

  async send(msg, text, opts = {}) {
    const { ai, config } = this.uniqueAiForChat(msg);

    var autoOpts = {}

    if (config.aiUsesReplies === "YES") {
      autoOpts.reply_to_message_id = msg.message_id;
      autoOpts.allow_sending_without_reply = true;
      autoOpts.reply_markup = JSON.stringify({ force_reply: msg.chat.type !== 'private', selective: true });

      if (config.aiDisableNotificationForReplies === "YES") {
        autoOpts.disable_notification = true;
      }
    }

    if (config.aiForceReplies === "YES") {
      autoOpts.reply_markup = JSON.stringify({ force_reply: true, selective: true });
    }

    if (config.aiDisableNotification === "YES") {
      autoOpts.disable_notification = true;
    }

    if (
      !autoOpts.hasOwnProperty('reply_to_message_id')
      && msg.message_thread_id
      && msg.chat.is_forum
    ) {
      autoOpts.message_thread_id = msg.message_thread_id;
    }

    autoOpts = { ...autoOpts, ...opts }

    let promise = null;
    const maxLen = 4096 - 8;

    const n = Math.ceil(text.length / maxLen);

    for (let i = 0; i < n; i++) {
      const chunk = text.substring(i * maxLen, i * maxLen + maxLen);

      if (!promise) {
        promise = this.bot.sendMessage(msg.chat.id, text.length < maxLen ? text : `${chunk}`, autoOpts);
      } else {
        await sleep(1000);
        promise = promise.then(msg => {
          return this.bot.sendMessage(msg.chat.id, `${chunk}`, autoOpts);
        });
      }
    }

    return promise.catch(ex => {
      debugOut(ex.message + ':\n' + JSON.stringify(msg, null, 2));
      throw ex;
    });
  }

  async completeResponseProbabilities(msg, uniqueAi) {
    const prompt = [
      'Rate a group chat messages probability in percent (0-100) and for each of the following statements:',
      '1. The message is directed to you',
      '2. The message is directed to somebody else',
      '3. The message is relevant',
      '4. You should respond to this message',
      '',
      'Reply with only a JSON object using the format: { directed_at_me: probability, directed_at_someone: probability, relevant: probability, should_respond: probability }',
      'If not enough information is available, set all probabilities to 0',
      '',
      'Message: {%message%}'
    ];

    return uniqueAi.createCompletion(prompt, { message: msg.text }).then(response => {
      uniqueAi.wipeMemory(prompt);

      return extractJSON(response.join('\n'));
    });
  }

  async completeMessageConditional(msg) {
    const { uniqueAi, config } = this.uniqueAiForChat(msg);

    if (!msg.text) { return }
    if (config.aiEnabled !== "YES") { return }
    if (
      msg.reply_to_message
      && this.aiIgnoreReply[msg.chat.id]
      && this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id]
      && (
        this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id] === msg.reply_to_message.message_id
        || (
          msg.reply_to_message.message_thread_id
          && this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id] === msg.reply_to_message.message_thread_id
        )
      )
    ) { return }

    if (config.aiAlwaysReply === "YES" || msg.text.includes(`@${this.botInfo.username}`)
      || (msg.reply_to_message && msg.reply_to_message.from.id === this.botInfo.id)) {
      return uniqueAi.createCompletion([msg.text], {});
    } else {
      return this.completeResponseProbabilities(msg, uniqueAi).then(rating => {
        if (rating.directed_at_me >= rating.directed_at_someone && rating.relevant >= 50) {
          return uniqueAi.createCompletion([msg.text], {});
        }
      });
    }
  }

  parseCommand(text) {
    if (text) {
      const regex = /^\/([^\s]+)\s?(.*)$/;
      const matches = text.match(regex);

      if (matches) {
        const commandName = matches[1].replace(`@${this.botInfo.username}`, '').toLowerCase();
        const params = matches[2] ? matches[2].split(' ') : [];
        return { commandName, params };
      }
    }

    return null;
  }

  async executeCommand(msg, name, params) {
    const command = this.commands.getByName(name);

    if (command) {
      if (command.ownerOnly && !this.isBotOwner(msg.from.id)) {
        return this.send(msg, 'You do not have permission to execute this command. 👮🏿');
      } else if (command.adminOnly && !this.isBotAdmin(msg.from.id)) {
        return this.send(msg, 'You do not have permission to execute this command. 👮🏻‍♂️');
      } else {
        return command.callback(msg, params);
      }
    }
  }

  async shutdown(msg) {
    debugOut('Bot shutting down...');

    try {
      Licensing.saveToStorage();
    } catch (ex) {
      debugOut(ex.message + ':\n' + JSON.stringify(this.bot, null, 2));
    }

    try {
      this.saveToStorage();
    } catch (ex) {
      debugOut(ex.message + ':\n' + JSON.stringify(this.bot, null, 2));
    }

    try {
      if (msg) {
        return this.send(msg, 'Halted').finally(() => {
          this.saveToStorage();
          this.bot.stopPolling().finally(() => {
            process.exit();
          });
        });
      } else {
        return this.bot.stopPolling().finally(() => {
          process.exit(0);
        });
      }
    } catch (error) {
      debugOut(error.message);
      process.exit(-1);
    }
  }

  main() {
    this.loadFromStorage();

    if (!this.botToken) {
      this.saveToStorage();
      throw new Error('No bot token available. Exiting...');
    }

    const bot = new TelegramBot(this.botToken, { polling: true });

    bot.getMe().then((botInfo) => {
      this.bot = bot;
      this.botInfo = botInfo;

      this.bot.on('message', (msg) => this.handleMessage(msg));
      this.bot.on('edited_message', (msg) => this.handleEditedMessage(msg));
      this.bot.on('edited_channel_post', (msg) => this.handleEditedMessage(msg));
      this.bot.on('polling_error', (error) => console.error(error));

      this.SetupContextCommands();

      debugOut('Bot is running...');
    });

    process.on('SIGTERM', () => {
      debugOut('!!! SIGTERM !!!');
      this.shutdown();
    });
    process.on('SIGINT', () => {
      debugOut('!!! SIGTERM !!!');
      this.shutdown();
    });
  }
}

module.exports = {
  CuteAiTelegramBot
};
