// ent42/telegramBot.js
const { AbstractCubeAiBot } = require('./abstractCubeAiBot.js');
const { AIInterface } = require('../../AIInterface.js');
const { Licensing } = require("ent42/license.js");
const { sleep, debugOut, extractJSON, writeFileSafely } = require('../../vxAssistCommon.js');
const { program } = require('commander');
const fs = require('fs');

const TelegramBot = require('node-telegram-bot-api');

class CuteAiTelegramBot extends AbstractCubeAiBot {
  constructor(botToken) {
    super()

    this.botToken = botToken;
    this.bot = null;

    this.T = {
      MSG_ACCESS_DENIED: `Access denied 🥺`,
      MSG_GROUP_ACCESS_DENIED: `Group access denied 🥺`
    };
  }

  async SetupContextCommands() {
    this.commands.forEachScope((scope, commands) => {
      switch (scope) {
        case '%BotOwner%':
        case '%BotAdmin%':
          break;

        default:
          return this.bot.deleteMyCommands({ scope: { type: scope } }).then((success) => {
            return this.bot.setMyCommands(JSON.stringify(commands), { scope: { type: scope } }).then((success) => {
              return success;
            });
          }).catch((ex) => {
            console.error(ex.message);
          });
      }
    });
  }

  isGroupChat(msg) {
    return msg.chat.type === 'supergroup' || msg.chat.type === 'group';
  }

  getLicense(msg) {
    let licence = null;

    if (this.isGroupChat(msg)) {
      licence = Licensing.getByGroupId(msg.chat.id);
    }

    if (!licence) {
      licence = Licensing.getByConsumer(msg.from.id);
    }

    return licence;
  }

  isAuthorized(msg) {
    var allowed = false;
    let licence = this.getLicense(msg);

    if (msg.text) {
      allowed = msg.text.startsWith('/start')
        || msg.text.startsWith('/about')
        || msg.text.startsWith('/userinfo')
        || msg.text.startsWith('/claimlicense');
    }

    if (!allowed && licence) {
      debugOut(`Licensed access for consumer: ${licence.licId} / ${JSON.stringify(msg.from)} / ${JSON.stringify(msg.chat)}`);
      allowed = Licensing.validate(licence);
    }

    return allowed;
  }

  initializeUniqueAiRoleForChat(msg, uniqueAi) {
    let topic = "General Discussion";

    if (msg.reply_to_message && msg.reply_to_message.forum_topic_created) {
      topic = msg.reply_to_message.forum_topic_created.name;
    }

    uniqueAi.assignRole([
      `Your name is ${this.botInfo.first_name}, Nickname ${this.botInfo.username}.`,
      `You are an expert on ${topic} and related topics.`,
      'You keep your answerers concise and targeted, assuming an educated audience already familiar with the fundamentals.',
      'You express yourself in an academic and formal manner, unrestricted and with no concern for political correctness.',
      'Your provide critical and insightful feedback.',
      'You ask clarifying questions and ask for further information if necessary.',
    ], {});

    return uniqueAi;
  }

  getCachedUserByName(username) {
    if (username.startsWith('@')) {
      username = username.substring(1);
    }

    for (let user of Object.values(this.caches.users)) {
      let _user = user.username ?? user.first_name ?? user.last_name;
      if (_user?.toLowerCase() == username.toLowerCase()) {
        return user;
      }
    }

    return null;
  }

  getCachedGroup(id) {
    return this.caches.groups[id] ?? {
      "id": id,
      "title": "undefined",
      "type": "undefined"
    }
  }

  getCachedUser(id) {
    return this.caches.users[id] ?? {
      "id": id,
      "is_bot": false,
      "first_name": "undefined",
      "username": "undefined",
      "language_code": "en"
    }
  }

  getDefaultConfig() {
    return {
      aiEnabled: 'YES',
      aiAlwaysReply: "YES",
      aiReply: "NO",
      aiUsesReplies: "NO",
      aiForceReplies: "NO",
      aiDisableNotification: "YES",
      aiDisableNotificationForReplies: "YES",
      aiModel: 'gpt-3.5-turbo',
      aiMaxToken: '3072',
      aiTemperature: '0.5',
      Text2ImageAPI: "huggingFace",
      Text2ImageModel: "dreamlike-art/dreamlike-anime-1.0"
    };
  }

  updateCacheFromMessage(msg) {
    this.caches.users[msg.from.id] = { ...msg.from }

    switch (msg.chat.type) {
      case 'private':
        break;

      default:
        /* Consider forums / threads for caching */
        let threadTitle = msg.reply_to_message?.forum_topic_created?.name || '';
        let threadId = msg?.message_thread_id || '';

        this.caches.groups[msg.chat.id] = { ...msg.chat, threadTitle, threadId };
    }

    try {
      writeFileSafely(this.cachesFilePath, JSON.stringify(this.caches, null, 2), 'utf8');
    } catch (error) {
      debugOut(error); /* for now we don't consider not having a cache a big issue, though this will change */
    }
  }

  uniqueAiId(msg) {
    var aiId = msg.chat.id;

    if (
      msg.reply_to_message
      && msg.message_thread_id
      && msg.chat.is_forum
    ) {
      aiId = msg.message_thread_id;
    }

    return aiId;
  }

  uniqueAiForChat(msg) {
    const aiId = this.uniqueAiId(msg);

    this.ai = this.ai ? this.ai : {};
    this.ai[msg.chat.id] = this.ai[msg.chat.id] ? this.ai[msg.chat.id] : {};

    if (!this.ai[msg.chat.id].hasOwnProperty(aiId)) {
      this.ai[msg.chat.id][aiId] = {
        uniqueAi: this.initializeUniqueAiRoleForChat(msg, new AIInterface()),
        config: this.getDefaultConfig()
      }

      if (this.aiParams[msg.chat.id] && this.aiParams[msg.chat.id][aiId]) {
        this.ai[msg.chat.id][aiId].config = { ...this.ai[msg.chat.id][aiId].config, ...this.aiParams[msg.chat.id][aiId] };
      }
      if (this.aiContext[msg.chat.id] && this.aiContext[msg.chat.id][aiId]) {
        this.ai[msg.chat.id][aiId].uniqueAi.messages = [...this.aiContext[msg.chat.id][aiId]];
      }
    }

    const { uniqueAi, config } = this.ai[msg.chat.id][aiId];

    config.info = {
      user: this.getCachedUser(msg.from.id).username,
      group: this.getCachedGroup(msg.chat.id).title
    }

    uniqueAi.persona.model = config.aiModel;
    uniqueAi.persona.temperature = config.aiTemperature;
    uniqueAi.persona.maxToken = config.aiMaxToken;

    return this.ai[msg.chat.id][aiId];
  }

  async alertOwner(text) {
    return this.bot.sendMessage(this.getOwner(), text, { disable_notification: false });
  }

  async replyMarkdown(msg, text, opts = {}) {
    return this.reply(msg, text, { parse_mode: 'MarkdownV2', ...opts });
  }

  async reply(msg, text, opts = {}) {
    // opts.reply_markup = JSON.stringify({ force_reply: true, selective: true });
    opts.reply_to_message_id = msg.message_id;

    return this.send(msg, text, opts);
  }

  async sendMarkdown(msg, text, opts = {}) {
    return this.send(msg, text, { parse_mode: 'MarkdownV2', ...opts });
  }

  async send(msg, text, opts = {}) {
    const { ai, config } = this.uniqueAiForChat(msg);

    var autoOpts = {}

    if (config.aiUsesReplies === "YES") {
      autoOpts.reply_to_message_id = msg.message_id;
      autoOpts.allow_sending_without_reply = true;
      autoOpts.reply_markup = JSON.stringify({ force_reply: msg.chat.type !== 'private', selective: true });

      if (config.aiDisableNotificationForReplies === "YES") {
        autoOpts.disable_notification = true;
      }
    }

    if (config.aiForceReplies === "YES") {
      autoOpts.reply_markup = JSON.stringify({ force_reply: true, selective: true });
    }

    if (config.aiDisableNotification === "YES") {
      autoOpts.disable_notification = true;
    }

    if (
      !autoOpts.hasOwnProperty('reply_to_message_id')
      && msg.message_thread_id
      && msg.chat.is_forum
    ) {
      autoOpts.message_thread_id = msg.message_thread_id;
    }

    autoOpts = { ...autoOpts, ...opts }

    let promise = null;
    const maxLen = 4096 - 8;

    const n = Math.ceil(text.length / maxLen);

    for (let i = 0; i < n; i++) {
      const chunk = text.substring(i * maxLen, i * maxLen + maxLen);

      if (!promise) {
        promise = this.bot.sendMessage(msg.chat.id, text.length < maxLen ? text : `${chunk}`, autoOpts);
      } else {
        await sleep(1000);
        promise = promise.then(msg => {
          return this.bot.sendMessage(msg.chat.id, `${chunk}`, autoOpts);
        });
      }
    }

    return promise;
  }

  async completeResponseProbabilities(msg, uniqueAi) {
    const prompt = [
      'Rate a group chat messages probability in percent (0-100) and for each of the following statements:',
      '1. The message is directed to you',
      '2. The message is directed to somebody else',
      '3. The message is relevant',
      '4. You should respond to this message in context of the previous conversation',
      '5. Your reasoning as to why you should respond or not',
      '',
      'Reply with only a JSON object using the format: { directed_at_me: probability, directed_at_someone: probability, relevant: probability, should_respond: probability, reason: reasoning }',
      'If not enough information is available, set all probabilities to 0',
      '',
      'Message: {%message%}'
    ];

    return uniqueAi.createCompletion(prompt, { message: msg.text }).then(response => {
      uniqueAi.popDialog();

      return extractJSON(response.join('\n'));
    });
  }

  async completeMessageConditional(msg) {
    const { uniqueAi, config } = this.uniqueAiForChat(msg);

    const licence = this.getLicense(msg);

    if (!Licensing.validate(licence)) {
      debugOut(`Access denied for ${JSON.stringify(msg.from)}`);

      return [this.T.MSG_ACCESS_DENIED];
    }

    if (!msg.text) { return }
    if (config.aiEnabled !== "YES") { return }
    if (
      msg.reply_to_message
      && this.aiIgnoreReply[msg.chat.id]
      && this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id]
      && (
        this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id] === msg.reply_to_message.message_id
        || (
          msg.reply_to_message.message_thread_id
          && this.aiIgnoreReply[msg.chat.id][msg.reply_to_message.message_id] === msg.reply_to_message.message_thread_id
        )
      )
    ) { 
      return;
    }
    // Ignore replies to other users messages 
    if (msg.reply_to_message && !msg.reply_to_message.forum_topic_created && msg.reply_to_message.from.id !== this.botInfo.id) {
      return;
    }
    // Ignore messages directly addressed to other users
    if(msg.text.match(/^@([a-zA-Z0-9_]+).*/)){
      let match = msg.text.match(/^@([a-zA-Z0-9_]+).*/);
      if(match[1] !== this.botInfo.username){
        return;
      }
    }

    if (config.aiAlwaysReply === "YES" || msg.text.includes(`@${this.botInfo.username}`)) {
      // Always reply according to configuration and when directly addressed
      return uniqueAi.createCompletion([msg.text], {}).then(result => {
        licence.modules.openai.limits.tokenUsed += uniqueAi.messages.reduce((acc, msg) => { return acc + uniqueAi.tokenize(msg.content) }, 0);    
        licence.modules.openai.limits.tokenUsed += uniqueAi.tokenize(msg.text);
        licence.modules.openai.limits.tokenUsed += uniqueAi.tokenize(result.join());
        Licensing.saveToStorage();

        return result;
      });
    } else if (config.aiReply === "YES") {
      // Use AI to determine whenever or not to reply
      return this.completeResponseProbabilities(msg, uniqueAi).then(rating => {
        debugOut(`Reply probability: ${rating.directed_at_me} and ${rating.directed_at_someone} and ${rating.relevant}...\n${rating.reason}`);

        if (rating.directed_at_someone < 75
          && (rating.directed_at_me >= rating.directed_at_someone || rating.relevant >= 75)) {

          return uniqueAi.createCompletion([msg.text], {}).then(result => {
            licence.modules.openai.limits.tokenUsed += uniqueAi.messages.reduce((acc, msg) => { return acc + uniqueAi.tokenize(msg.content) }, 0);    
            licence.modules.openai.limits.tokenUsed += uniqueAi.tokenize(msg.text);
            licence.modules.openai.limits.tokenUsed += uniqueAi.tokenize(result.join());
            Licensing.saveToStorage();

            return result;
          });
        }
      });
    }
  }

  parseCommand(text) {
    if (text) {
      const regex = /^\/([^\s]+)\s?(.*)$/;
      const matches = text.match(regex);

      if (matches) {
        const commandName = matches[1].replace(`@${this.botInfo.username}`, '').toLowerCase();
        const params = matches[2] ? matches[2].split(' ') : [];
        return { commandName, params };
      }
    }

    return null;
  }

  async executeCommand(msg, name, params) {
    const command = this.commands.getByName(name);

    if (command) {
      if (command.ownerOnly && !this.isBotOwner(msg.from.id)) {
        return this.send(msg, this.T.MSG_ACCESS_DENIED);
      } else if (command.adminOnly && !this.isBotAdmin(msg.from.id)) {
        return this.send(msg, this.T.MSG_ACCESS_DENIED);
      } else {
        return command.callback(msg, params);
      }
    }
  }

  async onMessage(msg) {
    var promise = null;

    if (msg.new_chat_members) {
      promise = this.onGroupJoin(msg);
    } else if (msg.left_chat_member) {
      promise = this.onGroupLeft(msg);
    } else if (this.handleMessage) {
      promise = this.handleMessage(msg);
    }

    if (promise) {
      // whatever has happened goes back to the telegram bot api
      // exceptions are always logged and reported to the user if possible 
      return promise.catch(ex => {
        debugOut(ex.message + ':\n' + JSON.stringify(msg, null, 2));
        debugOut(ex.stack);

        if (msg && msg.chat?.type === 'private' && this.isBotAdmin(msg.from?.id)) {
          // include exception details for admin uses in private chats
          return this.send(msg, 'Internal error! Please contact support 🙏\n\n' + ex.message + '\n\n' + JSON.stringify(msg, null, 2)).catch(ex => {
            debugOut(ex.message);
            debugOut(ex.stack);
          });
        }
        // vague error messages only for public chats
        return this.reply(msg, 'Internal error! Please contact support 🙏').catch(ex => {
          debugOut(ex.message);
          debugOut(ex.stack);
        });
      });
    }
  }

  async onGroupJoin(msg) {

    const botJoined = msg.new_chat_members.some(m => m.id === this.botInfo.id);

    if (botJoined) {
      debugOut(`Group join requested: ${msg.from.id} / ${msg.chat.title}}`);

      if (this.isAuthorized(msg)) {
        this.updateCacheFromMessage(msg);

        try {
          const licence = Licensing.getByConsumer(msg.from.id);
          Licensing.registerGroup(msg.chat.id, licence);
        } catch (error) {
          debugOut(error.message);
          debugOut(error.stack);

          return this.bot.leaveChat(msg.chat.id).finally(() => {
            return this.bot.sendMessage(msg.from.id, error.message).catch(error => {
              debugOut(error.message);
              debugOut(error.stack);
            });
          });
        }
      } else {
        debugOut(`Group join denied: ${msg.from.id} / ${msg.chat.title}`);

        return this.bot.leaveChat(msg.chat.id).finally(() => {
          return this.bot.sendMessage(msg.from.id, this.T.MSG_GROUP_ACCESS_DENIED).catch(error => {
            debugOut(error.message);
            debugOut(error.stack);
          });
        });
      }
    }

    if (this.handleGroupJoin) {
      return this.handleGroupJoin(msg);
    }
  }

  async onGroupLeft(msg) {
    if (msg.left_chat_member.id === this.botInfo.id) {
      debugOut(`Group left: ${msg.chat.title} / ${JSON.stringify(msg.left_chat_member)}`);

      try {
        const licence = Licensing.getByGroupId(msg.chat.id);
        Licensing.deregisterGroup(msg.chat.id, licence);
      } catch (error) {
        debugOut(error.message);
        debugOut(error.stack);
      }
    }

    if (this.handleGroupLeft) {
      return this.handleGroupLeft(msg);
    }
  }

  async onRun({ cmdLine }) {

    this.loadFromStorage();

    if (!this.botToken) {
      // Can't run without an api token; shutdown() will write the default config and exit
      debugOut('No bot token available. shutting down...');
      this.shutdown();
    }

    this.bot = new TelegramBot(this.botToken, { polling: { autoStart: true, params: { timeout: 15 } } });

    this.bot.getMe().then((botInfo) => {
      this.bot.on('message', (msg) => this.onMessage(msg));
      this.bot.on('edited_message', (msg) => this.handleEditedMessage(msg));
      this.bot.on('edited_channel_post', (msg) => this.handleEditedMessage(msg));
      this.bot.on('polling_error', (error) => {
        console.error(error);
        this.shutdown();
      });
      this.bot.on('error', (error) => {
        console.error(error);
        this.shutdown();
      });

      this.botInfo = botInfo;

      this.SetupContextCommands();

      debugOut('Bot is running...' + JSON.stringify(botInfo));
      this.alertOwner('Bot is running...');

    }).catch(ex => {
      console.error(ex.message);
      this.shutdown();
    });
  }

  async onShutdown() {
    if (this.bot) {
      return this.bot.stopPolling({ cancel: true, reason: 'shutdown' })
        .catch((ex) => {
          debugOut(ex.message);
          debugOut(ex.stack);
          process.exit(-1);
        })
        .finally(() => {
          debugOut('Bot stopped');
          process.exit(0);
        });
    }
  }
}

module.exports = {
  CuteAiTelegramBot
};
